ultravox/tools/ds_tool/ds_tool.py

```diff
@@ -223,6 +228,124 @@ def __post_init__(self):
+            split_name, ds_split, 0, len(ds_split)
+        )
+
+        while len(failed_chunk_ranges) > 0:
+            new_failed_ranges = []
+            for start, end in failed_chunk_ranges:
+                print(f"Retrying failed chunk range [{start}, {end})")
+                new_failed_ranges.extend(
+                    self._dynamic_chunk(split_name, ds_split, start, end)
+                )
+            failed_chunk_ranges = new_failed_ranges
+        print(f"Could not upload chunks: {self.chunks_not_uploaded}")
+        print(
+            f"Finished processing and uploading all chunks for split {split_name}. Total samples processed: {self.total_samples_processed}"
+        )
+
+    def _dynamic_chunk(
+        self,
+        split_name: str,
+        ds_split: datasets.Dataset,
+        start_index: int,
+        end_index: int,
+    ):
+        original_chunk_size = end_index - start_index
+        if original_chunk_size < self.args.chunk_split_threshold:
+            total_chunks = 1
+            chunk_size = original_chunk_size
+            print(
+                f"Chunk range [{start_index}, {end_index}) is too small to split further. Processing and uploading as a single chunk."
+            )
+        else:
+            total_chunks = self.args.max_chunk_split
+            chunk_size = math.ceil(original_chunk_size / total_chunks)
+        failed_chunk_ranges = []
+        print(
+            f"Processing and uploading {total_chunks} chunks for range [{start_index}, {end_index}) with chunk size {chunk_size}"
+        )
+        for i in range(start_index, end_index, chunk_size):
+            chunk_start = i
+            chunk_end = min(i + chunk_size, start_index + original_chunk_size)
+
+            ds_chunk = ds_split.select(range(chunk_start, chunk_end))
+            ds_chunk_name = f"chunk-range-{chunk_start:09d}-{chunk_end:09d}.parquet"
+            ds_chunk_hub_path = os.path.join(
+                str(self.args.upload_subset), split_name, ds_chunk_name
+            )
+            ds_chunk_cache_path = os.path.join(
+                self.cache_dir,
+                self.args.dataset_name.replace("/", "__"),
+                str(self.args.upload_subset),
+                split_name,
+                ds_chunk_name,
+            )
+            try:
+                if os.path.exists(ds_chunk_cache_path):
+                    print(
+                        f"Skipping chunk {ds_chunk_name} as it has already been processed and uploaded."
+                    )
+                    ds_chunk_processed = datasets.Dataset.from_parquet(
+                        ds_chunk_cache_path
+                    )
+                else:
+                    print(f"Processing chunk {ds_chunk_name}")
+                    ds_chunk_processed = self._process(ds_chunk)
+                    self._upload(ds_chunk_processed, ds_chunk_hub_path, split_name)
+                    ds_chunk_processed.to_parquet(ds_chunk_cache_path)
```

Comment:
Won't you loose the chunk if upload fails? I suggest switching the order:
```suggestion
                    ds_chunk_processed.to_parquet(ds_chunk_cache_path)
                    self._upload(ds_chunk_processed, ds_chunk_hub_path, split_name)
```

Reply:
the idea is if the upload fails, we then want to break it up into smaller chunks to upload. We actually don't really need to save the chunk here, because all we're using the saved chunk for is to check if that chunk has already been uploaded to hf. I had some other ideas before about retrying saved chunks, but figured splitting into smaller chunks would be simpler, and also handles more edge cases.

Reply:
Hmm, that seems wasteful if you just happen to get a networking issue. But yeah I guess you mentioned sometimes these issues are persistent for some reason, is that true?

Then I suggest two things:
1. add a comment to explain why this decision was made and why switching the order will break the code. I'd make sure that the assumption "saved chunks indicate what's been uploaded to HF" is quite prominent.

2. (not required for this PR) try to add a retry for `_upload` to handle transient network failures

Reply:
So I'm not actually loading the cached parquet anywhere. I just use it to check if the path exists and if it does then we skip processing AND uploading. So its important that it comes before the upload, or else we potentially end up in a case where the chunk is processed and cached, but not uploaded. The next time we run the program, the program skips over the process and upload step if it finds the chunk in the cache. Reason why I tie the process and upload step together is because I've noticed that reprocessing can sometimes help with uploading errors. Maybe its best to not store the whole parquet here, rather just a note in the cache that the chunk has already been uploaded to hf, but then again having the whole parquet maybe useful if something unexpected happens.